### ISSUE TITLE
UMat memory leak on Intel GPUs when using multiple threads

### ISSUE THREAD

**Issue by leugenea** [opencv/opencv#10101]
##### System information (version)

- OpenCV => 3.3.1
- Operating System / Platform => Windows 10 64 Bit
- Compiler => Visual Studio 2013

##### Detailed description

Using UMat with OpenCL on Intel videocards in multiple host threads causes GPU memory leak.

##### Steps to reproduce

1. Create function that: loads image into `cv::Mat`, copies it to `cv::UMat` using `copyTo` method, runs some OpenCL-related function like CascadeClassifier's `detectMultiScale` or `cv::equalizeHist`, and returns.
2. Create loop: create thread, run function from 1 in this thread, join thread.
3. Watch GPU memory increasing indefinitely.

![intel_gpu_leak](https://user-images.githubusercontent.com/534699/32901516-09481228-cb01-11e7-93e7-725f5ed2aa9a.png)

##### Additional notes

1. If not using Intel GPU (tested on various NVidia GPUs), there's no memory leak.
2. If using OpenCL from fixed number of threads (for example, in fixed-size threadpool), there's no memory leak.
3. Even if I release UMat's memory using `release()` method, memory will leak.

**Comment by alalek** [comment]
Consider to use ThreadPools instead of spawning new threads.

**Comment by leugenea** [comment]
Yes, of course. But isn't it a bit strange that OpenCL memory isn't deallocated until program terminates?
